WIP: sRGB Loader Test and Fixes for multiple image formats

Florastamine opened the pull request:
General:
- [x] ~Replace legacy color converters (JPG/PNG/TGA)~
- [x] ~Handle incorrect flipping of image buffers (JPG/PNG/TGA)~
- [x] ~Handle unsupported color formats during loading/saving (JPG/PNG/TGA)~
- [x] ~More test images (preferably PD/commonly used images - Lenny, jelly beans, parrot, etc.)~
- [x] ~Convert formats for display testing as needed~ (Superseded in #305)
- [x] ~Fix Linux-related build errors~

**JPG**
- [x] ~Handle grayscale (`JCS_GRAYSCALE` would load the image as `EF_R8_SRGB`)~

**PNG**
- [x] ~Fix 8-bit grayscale (8-bit got expanded to RGB(A)) & 32-bit RGBA~

**TGA**
- [x] ~Fix 8-bit grayscale & 32-bit RGBA (+RLE)~
- [x] ~Correct handling of TGA headers parsing~

Commit:
Change the code to read the test list from an external file

Commit:
Don't hardcode testlist.txt

Commit:
Revert "Don't hardcode testlist.txt"

This reverts commit a8ccadbf87a4ef0d3c2686582ee64a84a53cf54d.

Commit:
Minor refactoring + allow Targa formats to load as normal

Commit:
Fixed JPG writer

Commit:
Write grayscale JPG

devshgraphicsprogramming commented on src/irr/asset/CImageWriterJPG.cpp:
```diff
@@ -105,6 +105,7 @@ static bool writeJPEGFile(io::IWriteFile* file, const asset::CImageData* image,
 	{
         case asset::EF_R8G8B8_UNORM:
+		case asset::EF_R8G8B8_SRGB:
 			format = video::CColorConverter::convert_R8G8B8toR8G8B8;
```
We want to get rid of CColorConverter, it is legacy code from old irrlicht and doesn't have good format coverage/support.

Maybe you can use the `decode` and `encode` functions that @Crisspl  has made.

JPEG only supports SRGB without conversion (natively), if you encounter `EF_R8G8B8_UNORM` or anything similar then you need a convert to `EF_R8G8B8_SRB` from the input format.

However you have my approval to error-out in such case.

devshgraphicsprogramming commented on src/irr/asset/CImageWriterJPG.cpp:
```diff
@@ -102,6 +102,8 @@ static void jpeg_file_dest(j_compress_ptr cinfo, io::IWriteFile* file)
+	
 	switch( image->getColorFormat () )
 	{
         case asset::EF_R8G8B8_UNORM:
```
none of the below code with CColorConverter will work, that conversion code is legacy

Commit:
Use video::convertColor() for the JPEG writer

Commit:
Convert EF_R8G8B8_UNORM

Commit:
Fixed JPG writer, fixed warning + bump the default quality a bit

Commit:
Replaced color converters for TGA loaders + flipping; dropped 8bit nongrayscale

Commit:
Fixed the TGA writers

Commit:
Fixed PNG loader & writer

devshgraphicsprogramming commented on src/irr/asset/CImageWriterJPG.cpp:
```diff
@@ -209,7 +209,7 @@ bool CImageWriterJPG::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
     const asset::E_WRITER_FLAGS flags = _override->getAssetWritingFlags(ctx, image, 0u);
     const float comprLvl = _override->getAssetCompressionLevel(ctx, image, 0u);
-	return writeJPEGFile(file, image, (!!(flags & asset::EWF_COMPRESSED))*(1.f-comprLvl)*100.f); // if quality==0, then it defaults to 75
+	return writeJPEGFile(file, image, (!!(flags & asset::EWF_COMPRESSED)) * (uint32_t)((1.f-comprLvl)*100.f)); // if quality==0, then it defaults to 75
```
don't use C-style cast, use `static_cast<uint32_t>`

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderTGA.cpp:
```diff
@@ -84,6 +84,29 @@ bool CImageLoaderTGA::isALoadableFileFormat(io::IReadFile* _file) const
 
+// convertColorFlip() does color conversion as well as taking care of properly flipping the given image.
+template <typename T, E_FORMAT srcFormat, E_FORMAT destFormat>
+static void convertColorFlip(const T *src, T *dest, core::vector3d<uint32_t> &size, int c, bool flip)
```
this is the kind of thing that would be **very** useful in top level header, maybe in `./include/irr/video/convertColor.h`

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderTGA.cpp:
```diff
@@ -84,6 +84,29 @@ bool CImageLoaderTGA::isALoadableFileFormat(io::IReadFile* _file) const
+{
+	T *in = (T *) src;
+	T *out = (T *) dest;
+	uint32_t stride = size.X * c;
```
your strides will be different for non-identical src and dest Formats

moreover images usually have an issue with row-pitch alignment (for cache, processor architecture reasons 8 byte alignment works best, so you cannot really have 8bit images of arbitrary width, they get a padding)

devshgraphicsprogramming replied:
just pass src and dest row pitches, instead of `c`

manhnt9 replied:
@devshgraphicsprogramming @Florastamine: is this solved?

devshgraphicsprogramming replied:
no idea where that function is now

devshgraphicsprogramming replied:
@Florastamine?

devshgraphicsprogramming replied:
still not fixed, using same stride for both the `in` and `out` pointer

manhnt9 replied:
@Florastamine

devshgraphicsprogramming replied:
> It doesn't do CImageData-to-CImageData type of conversion, hence only 1 stride

If you expect those kind of things, then you need to place `assert`s in the code

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderTGA.cpp:
```diff
@@ -165,54 +189,52 @@ asset::IAsset* CImageLoaderTGA::loadAsset(io::IReadFile* _file, const asset::IAs
+					return nullptr;
 				}
+				
+				image = new asset::CImageData(NULL,nullOffset,imageSize,0,asset::EF_R8_UNORM);
```
make sure that TGA does not mandate by the spec for greyscale images to also be gamma-encoded

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderPNG.cpp:
```diff
@@ -152,31 +152,30 @@ asset::IAsset* CImageLoaderPng::loadAsset(io::IReadFile* _file, const asset::IAs
+		switch (ColorType) {
+			case PNG_COLOR_TYPE_GRAY:
+			case PNG_COLOR_TYPE_GRAY_ALPHA:
+				png_set_expand_gray_1_2_4_to_8(png_ptr);
```
don't expand grey or grey-alpha to rgb or rgba

Commit:
Fixed texture flipping for JPG

Commit:
Fixed flipping for PNG

Commit:
Fixed flipping for TGA + fixed incorrect stride

Commit:
Read/write TGA gamma data

Commit:
Parentheses

devshgraphicsprogramming commented on src/irr/asset/CImageLoaderTGA.cpp:
```diff
@@ -77,9 +77,21 @@ bool CImageLoaderTGA::isALoadableFileFormat(io::IReadFile* _file) const
+		_file->read(&extension, sizeof(STGAExtensionArea));
+		
+		// TODO: Pass this to new asset::CImageData()
+		float gamma = extension.Gamma;
```
AFAIK if gamma is all zero then it means it also not present

devshgraphicsprogramming commented on src/irr/asset/CImageWriterTGA.cpp:
```diff
@@ -159,7 +159,7 @@ bool CImageWriterTGA::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
 	STGAExtensionArea extension;
 	extension.ExtensionSize = sizeof(extension);
-	extension.Gamma = format == asset::EF_R8_UNORM ? 1.0f : ((100.0f / 30.0f) - 1.0f);
+	extension.Gamma = (format == asset::EF_R8_UNORM) ? 1.0f : ((100.0f / 30.0f) - 1.0f);
```
can you explain the second gamma choice? why does it work out to 2.33333 ?

IEC 61966-2-1 works out more closely to 2.23333

devshgraphicsprogramming replied:
also, format==EF_R8_UNORM is not a check for gamma

devshgraphicsprogramming replied:
we have this beauty for this
https://github.com/buildaworldnet/IrrlichtBAW/blob/315172ea3d1b9b51da4c1ef30412c02b99f1066a/include/irr/asset/EFormat.h#L1085

devshgraphicsprogramming replied:
there should be a runtime counterpart to this, if not create one

Commit:
TGA related fixes

Commit:
Don't do save operations on unwriteable formats (DDS, BMP); allows notation on the testfile

devshgraphicsprogramming commented on examples_tests/09.ColorSpaceTest/main.cpp:
```diff
@@ -96,24 +96,32 @@ bool ReadLine(io::IReadFile* file, std::string &line)
 		filename += "."; filename += extension;
-
+		
+		bool writeable = (extension != "dds") && (extension != "bmp");
```
so you're nuking the BMP writer, correct?

Commit:
DDS: Refactoring this mess + removed dead code

Commit:
Switched to use EF_R8_SRGB

Commit:
NULLPTR EVERYTHANG

devshgraphicsprogramming commented on src/irr/asset/CImageWriterJPG.cpp:
```diff
@@ -155,6 +155,7 @@ static bool writeJPEGFile(io::IWriteFile* file, const asset::CImageData* image,
 					video::convertColor<EF_A1R5G5B5_UNORM_PACK16, EF_R8G8B8_SRGB>(src_container, dest, 1, dim.X, dim);
 					break;
+				case asset::EF_R8_SRGB:
 				case asset::EF_R8_UNORM:
```
you cannot write it out with memcpy if its not SRGB! conversion needed

devshgraphicsprogramming commented on src/irr/asset/CImageWriterPNG.cpp:
```diff
@@ -169,6 +171,7 @@ bool CImageWriterPNG::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
 	{
 		switch (format) {
+			case asset::EF_R8_SRGB:
 			case asset::EF_R8_UNORM:
```
CONVERSION NEEDED, OR ERROR OUT

devshgraphicsprogramming commented on src/irr/asset/CImageWriterTGA.cpp:
```diff
@@ -85,6 +81,7 @@ bool CImageWriterTGA::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
 			}
 			break;
+		case asset::EF_R8_SRGB:
 		case asset::EF_R8_UNORM:
```
convert or error out

devshgraphicsprogramming commented on src/irr/asset/CImageWriterTGA.cpp:
```diff
@@ -122,6 +119,7 @@ bool CImageWriterTGA::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
 		switch (format) {
 			// Do a pass-through, byte-wise copy for everything else except EF_R8G8B8(A8)_SRGB, which would need swizzling to EF_B8G8R8(A8)_SRGB.
+			case asset::EF_R8_SRGB:
 			case asset::EF_R8_UNORM:
```
convert or error out

devshgraphicsprogramming committed:
Update Readme.md

draft, needs a refactor

devshgraphicsprogramming committed:
Split the rules

devshgraphicsprogramming committed:
Merge pull request #300 from buildaworldnet/extension_rules

Updated Extension Rules

Commit:
Proper format conversion for EF_R8_UNORM

Commit:
Added testing suite

Commit:
Added 32-bit RLE TGAs

Commit:
Correct handling of TGA header parsing

Commit:
Replaced 8-bit TGA test image with the one from npedotnet

Commit:
Fixed Linux build errors

devshgraphicsprogramming committed:
Update Ext Readme

About a nice CMake script for the sRGB branch that can define an irr::ext:: library for any extension

Commit:
Bails out with error message on unknown colorspace

Commit:
Merge branch 'master' of https://github.com/buildaworldnet/IrrlichtBAW into sRGB

Commit:
Helpers to convert linear -> sRGB

Commit:
Fixed incorrect line reading on certain platforms where CR/LF bytes are inconsistent.

Commit:
Missing #include

Commit:
SColorf::fromSRGB(), SColorf::toSRGB()

devshgraphicsprogramming commented on include/SColor.h:
```diff
@@ -416,6 +417,22 @@ namespace video
+		inline static SColorf fromSRGB(SColorf&& input)
+		{
+			float color[3] = {input.r, input.g, input.b};
+			impl::SRGB2lin<float>(color);
```
you could just pass `reinterpret_cast<float[3]>(input.pointer)` or `reinterpret_cast<float[3]>(&input.r)` instead

devshgraphicsprogramming replied:
then an `std::move()` of the `input` into the `SColorf` constructor, like this
```cpp
return SColorf(std::move(input));
```

devshgraphicsprogramming replied:
ah, I forgot what the syntax for this is
https://stackoverflow.com/questions/17258749/is-there-a-way-of-casting-a-pointer-to-an-array-type

devshgraphicsprogramming commented on include/irr/video/decodePixels.h:
```diff
@@ -1283,6 +1286,17 @@ namespace irr { namespace video
         }
+		
+        template<>
+        inline void lin2SRGB<double>(double _lin[3])
```
@Crisspl  has this very function in `encodePixels.h` I believe?

devshgraphicsprogramming replied:
you can make his templated, and slam an assert about the template type (onyl float or double) in there

devshgraphicsprogramming commented on include/irr/video/decodePixels.h:
```diff
@@ -1293,6 +1307,16 @@ namespace irr { namespace video
                 lin[i] = s[i] * 255.;
         }
+
+        template<typename T>
```
templating is OK, but you need a `static_assert` which ensures that `T` can only be `float` or `double`

devshgraphicsprogramming replied:
@Florastamine, we need a static_assert, dont want someone using that with ints or vector classes

Commit:
sad violin intensifies
